### PR TITLE
[consensus] make Genesis an explicit enum entry

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

It's no-op in logics. Today we fake some fields in genesis, it'll get worse when we have reconfiguration genesis, so this is the first step to make it explicit what is a genesis block.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Unit tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 1]
```diff
@@ -195,7 +199,11 @@ impl<T: Payload> BlockStore<T> {
         };
         let compute_res = self
             .state_computer
-            .compute(parent_id, block.id(), block.get_payload())
+            .compute(
+                parent_id,
+                block.id(),
+                block.payload().unwrap_or(&T::default()),
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: Genesis will never be computed, here is the NIL block payload which we prepare to still have block prologue for it.

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 2]
```diff
@@ -119,7 +119,11 @@ impl<T: Payload> BlockStore<T> {
             .collect::<HashMap<_, _>>();
         for block in blocks {
             let compute_res = state_computer
-                .compute(block.parent_id(), block.id(), block.get_payload())
+                .compute(
+                    block.parent_id(),
+                    block.id(),
+                    block.payload().unwrap_or(&T::default()),
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: genesis is the root which won't be passed to this, I can simply add assertion here.

File: consensus/src/chained_bft/event_processor.rs [hunk 3]
```diff
@@ -802,20 +802,22 @@ impl<T: Payload> EventProcessor<T> {
...
-                .txn_manager
-                .commit_txns(
-                    committed.get_payload(),
-                    compute_result.as_ref(),
-                    committed.timestamp_usecs(),
-                )
-                .await
-            {
-                error!("Failed to notify mempool: {:?}", e);
+            if let Some(payload) = committed.payload() {
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/chained_bft/event_processor.rs]: NIL block doesn't have payload too and we shouldn't notify mempool about it

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 4]
```diff
@@ -121,8 +121,13 @@ impl<T: Payload> BlockStore<T> {
             .map(|qc| (qc.certified_block_id(), qc))
             .collect::<HashMap<_, _>>();
         for block in blocks {
+            assert!(!block.is_genesis_block());
```
Comment (aching) [inline consensus/src/chained_bft/block_storage/block_store.rs]:
```suggestion
            assert!(!block.is_genesis_block(), "A genesis block cannot be inserted into the BlockTree");
```

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (aching) [reply consensus/src/chained_bft/block_storage/block_store.rs]: Please add the assert as done above.

**Pull request closed without merging**